src/couch_index/test/couch_index_ddoc_updated_tests.erl

```diff
@@ -0,0 +1,134 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(couch_index_ddoc_updated_tests).
+
+-include_lib("couch/include/couch_eunit.hrl").
+-include_lib("couch/include/couch_db.hrl").
+
+setup() ->
+    DbName = ?tempdb(),
+    fake_mem3(DbName),
+    fake_index(),
+
+    DbNames = [mem3:name(Sh) || Sh <- mem3:local_shards(mem3:dbname(DbName))],
+    [Db1| RestDbs] = lists:map(fun(DBName) ->
+        {ok, DB} = couch_db:create(DBName, [?ADMIN_CTX]),
+        DB
+    end, DbNames),
+
+    % create a DDoc on the 1st Db
+    DDocID = <<"idx_name">>,
+    DDocJson = couch_doc:from_json_obj({[
+        {<<"_id">>, DDocID},
+        {<<"value">>, 1}
+    ]}),
+    {ok, _Rev} = couch_db:update_doc(Db1, DDocJson, []),
+    {ok, Db} = couch_db:reopen(Db1),
+    {ok, DDoc} = couch_db:open_doc(Db, DDocID, [ejson_body, ?ADMIN_CTX]),
+    Dbs = [Db | RestDbs],
+    {Dbs, DDoc}.
+
+
+teardown({Dbs, _DDoc}) ->
+    lists:foreach(fun(Db) ->
+        couch_db:close(Db),
+        couch_server:delete(Db#db.name, [?ADMIN_CTX])
+    end, Dbs),
+    (catch meck:unload(mem3)),
+    (catch meck:unload(test_index)),
+    ok.
+
+
+ddoc_update_test_() ->
+    {
+        "Check ddoc update actions",
+        {
+            setup,
+            fun() -> test_util:start_couch([]) end, fun test_util:stop_couch/1,
+            {
+                foreach,
+                fun setup/0, fun teardown/1,
+                [
+                    fun ddoc_update/1
+                ]
+            }
+        }
+    }.
+
+
+ddoc_update({[Db | _] = Dbs, DDoc}) ->
+    ?_test(begin
+        N = length(Dbs),
+        DDocID = DDoc#doc.id,
+
+        % run couch_index process for each Db
+        lists:foreach(fun(DB) ->
+            couch_index_server:get_index(test_index, DB, DDoc)
+        end, Dbs),
+        IndexesBefore = ets:match_object(
+            couchdb_indexes_by_db, {'$1', {DDocID, '$2'}}),
+        ?assertEqual(N, length(IndexesBefore)),
+
+        % update ddoc
+        DDocJson2 = couch_doc:from_json_obj({[
+            {<<"_id">>, DDocID},
+            {<<"value">>, 2},
+            {<<"_rev">>, couch_doc:rev_to_str(DDoc#doc.revs)}
+        ]}),
+        {ok, _} = couch_db:update_doc(Db, DDocJson2, []),
+
+        % assert that all index processes exited after ddoc updated
+        couch_index_server:handle_db_event(
+            Db#db.name, {ddoc_updated, DDocID}, {st, ""}),
+        timer:sleep(1000),
```

Comment:
+1 for avoiding the sleep call. There's been a lot of work lately trying to track down broken tests in CI on ASF Jenkins and Travis-CI because those build hosts are so slow that our timing stuff goes out the window.